src/erlcloud_ssm.erl

```diff
@@ -0,0 +1,589 @@
+    Opts1 = lists:ukeysort(1, proplists:unfold(Opts)),
+    lists:foldl(fun(Opt, A) -> opt_folder(Table, Opt, A) end, {[], []}, Opts1);
+opts(_, _) ->
+    error({erlcloud_ssm, opts_not_list}).
+
+%%%------------------------------------------------------------------------------
+%% Shared Decoders
+%%%------------------------------------------------------------------------------
+
+-type decode_opt() :: {typed, boolean()}.
+-type decode_opts() :: [decode_opt()].
+-type record_desc() :: {tuple(), field_table()}.
+
+-spec id(X) -> X.
+id(X) -> X.
+
+-spec id(X, decode_opts()) -> X.
+id(X, _) -> X.
+
+-type field_table() :: [{binary(), pos_integer(), 
+                         fun((jsx:json_term(), decode_opts()) -> term())}].
+
+-spec decode_folder(field_table(), json_pair(), decode_opts(), tuple()) -> tuple().
+decode_folder(Table, {Key, Value}, Opts, A) ->
+    case lists:keyfind(Key, 1, Table) of
+        {Key, Index, ValueFun} ->
+            setelement(Index, A, ValueFun(Value, Opts));
+        false ->
+            A
+    end.
+
+
+-spec decode_record(record_desc(), jsx:json_term(), decode_opts()) -> tuple().
+decode_record({Record, _}, [{}], _) ->
+    %% jsx returns [{}] for empty objects
+    Record;
+decode_record({Record, Table}, Json, Opts) ->
+    lists:foldl(fun(Pair, A) -> decode_folder(Table, Pair, Opts, A) end, Record, Json).
+
+%%%------------------------------------------------------------------------------
+%% Output
+%%%------------------------------------------------------------------------------
+-spec out(json_return(), decode_fun(), ssm_opts())
+         -> {ok, jsx:json_term() | tuple()} |
+            {simple, term()} |
+            {error, term()}.
+out({error, Reason}, _, _) ->
+    {error, Reason};
+out({ok, Json}, Decode, Opts) ->
+    case proplists:get_value(out, Opts, json) of
+        json ->
+            {ok, Json};
+        record ->
+            {ok, Decode(Json, [])};
+        map ->
+            {ok, erlcloud_util:proplists_to_map(Json)}
+    end.
+
+%%%------------------------------------------------------------------------------
+%% Shared Records
+%%%------------------------------------------------------------------------------
+
+-spec parameter_record() -> record_desc().
+parameter_record() ->
+    {#ssm_parameter{},
+        [
+            {<<"ARN">>, #ssm_parameter.arn, fun id/2},
+            {<<"DataType">>, #ssm_parameter.data_type, fun id/2},
+            {<<"LastModifiedDate">>, #ssm_parameter.last_modified_date, fun id/2},
+            {<<"Name">>, #ssm_parameter.name, fun id/2},
+            {<<"Selector">>, #ssm_parameter.selector, fun id/2},
+            {<<"SourceResult">>, #ssm_parameter.source_result, fun id/2},
+            {<<"Type">>, #ssm_parameter.type, fun id/2},
+            {<<"Value">>, #ssm_parameter.value, fun id/2},
+            {<<"Version">>, #ssm_parameter.version, fun id/2}
+        ]
+    }.
+
+-spec put_parameter_record() -> record_desc().
+put_parameter_record() ->
```

Comment:
Why does this action record exist but not one for the other actions you've implemented that have a response body, like GetParameter, GetParameters?

Reply:
Thanks for pointing that out. It turns out because I did not have one specifically for `GetParameters` when you use the `{out, record}` the `InvalidParameters` output gets hidden. I'm going to update this to include a record for this output only, I don't think that is required for `GetParameter` or `GetParametersByPath` though since they can simply be `#ssm_parameter{}` and `[#ssm_parameter{}]`. I defer to you and @nalundgaard though. Updated the PR with a `#ssm_get_parameters{}` record that will look like this:

```erlang
> erlcloud_ssm:get_parameters([{names, [<<"/root/test">>, <<"invalid_param">>]}, {out, record}]).
{ok,#ssm_get_parameters{invalid_parameters = [<<"invalid_param">>],
                        parameters = [#ssm_parameter{arn = <<"arn:aws:ssm:us-west-2:00000000:parameter/root/test">>,
                                                     data_type = <<"text">>,last_modified_date = 1612970194.057,
                                                     name = <<"/root/test">>,selector = undefined,
                                                     source_result = undefined,type = <<"String">>,
                                                     value = <<"myP@ssw0rd">>,version = 1}]}}
```

I also updated the unit to cover this